presto-main/src/main/java/com/facebook/presto/util/PrestoIterators.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.util;
+
+import com.google.common.collect.AbstractIterator;
+
+import java.util.Iterator;
+
+import static java.util.Objects.requireNonNull;
+
+public class PrestoIterators
+{
+    private PrestoIterators()
+    {
+    }
+
+    public static <T> Iterator<T> closeWhenExhausted(Iterator<T> iterator, AutoCloseable resource)
+    {
+        requireNonNull(iterator, "iterator is null");
+        requireNonNull(resource, "resource is null");
+
+        return new AbstractIterator<T>()
+        {
+            @Override
+            protected T computeNext()
+            {
+                if (iterator.hasNext()) {
+                    return iterator.next();
+                }
+                else {
+                    try {
+                        resource.close();
+                    }
+                    catch (Exception e) {
+                        throw new RuntimeException(e);
+                    }
+                    return endOfData();
```

Comment:
`endOfData` would not be called if exception is thrown. I suggest you rewrite this `else` clause as:
```java
                    endOfData();
                    try {
                        resource.close();
                        return null;
                    }
                    catch (Exception e) {
                        throw new RuntimeException(e);
                    }

```